### PR TITLE
Plugin framework

### PR DESCRIPTION
**Pull request by davidmckenzie**
Solves #130 

**No merge - still WIP, PR for visibility only**

### CODE REVIEW

File: server/db.js [hunk 1]
```diff
@@ -85,19 +75,135 @@ function init(release) {
...
                                     db.run("ALTER TABLE capcodes ADD telechat TEXT", function(err){ /* ignore error */ });
                                     db.run("ALTER TABLE capcodes ADD ignore INTEGER DEFAULT 0", function (err) { /* ignore error */ });
                                     db.run("ALTER TABLE capcodes ADD twitter INTEGER DEFAULT 0", function (err) { /* ignore error */ });
                                     db.run("ALTER TABLE capcodes ADD twitterhashtag TEXT", function (err) { /* ignore error */ });
                                     db.run("ALTER TABLE capcodes ADD discord INTEGER DEFAULT 0", function (err) { /* ignore error */ });
                                     db.run("ALTER TABLE capcodes ADD discwebhook TEXT", function (err) { /* ignore error */ });
-                                    db.run("PRAGMA user_version = "+release, function(err){ /* ignore error */ });
+                                    db.run("ALTER TABLE capcodes ADD pluginconf TEXT", function (err) { /* ignore error */ });
+                                    // begin scary stuff, consider hiding behind a solid object during this bit
```
Comment (DanrwAU) [inline server/db.js]: Should probably add a check here to ensure this only runs once. 
I used if (res..user_version < '20181118') when playing with Knex

Comment (davidmckenzie) [reply server/db.js]: Good catch - will make it so it only executes if upgrading from version prior to this release :)

Comment (davidmckenzie) [reply server/db.js]: Updated the upgrade scripts, will make this smarter at some point, if not made obsolete by the knex stuff

**Comment by DanrwAU** [comment]
Looks like some of the logging is still grabbing objects. 

Seeing this when loading the global settings page
```
3|pagermon  | 2018-12-03 13:02 +00:00: [pmon]  2018-12-04 00:12:12  debug : [object Object]
3|pagermon  | 2018-12-03 13:02 +00:00: [pmon]  2018-12-04 00:12:13  debug : [object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]
```

I'll test everything else in the morning now that i've got my test environment setup again and see if i spot anything else 👍 😉

**Comment by marshyonline** [comment]
Top work!
Updated well from a decently old version. plugins work as intended.

One suggestion i will put to the table tho - to make the httpd log a true access log and to make life simple for parsing it with something like awstats or goaccess - can we adjust line #42 of log.js to be just the following? I also this if we do this - we could skip it from showing in the console?
```info => `${info.message}` ```

I also noted when testing on windows - I had to create the logs directory before logs would be saved.

**Comment by davidmckenzie** [comment]
@marshyonline pushed changes there - the logs dir should have come across but looks like a gitignore was killing it :) for HTTP logging, good call there, what I've done is made it so that the HTTP things come in as debug level logs, and set the file to log debug regardless of log settings. That way if you set to debug level then you still see HTTP requests in CLI.

@DanrwAU bah, looks like some nastiness with winston v3 on object logging... will do a bit more digging there

**Comment by davidmckenzie** [comment]
Ok so this kinda sucks but it'll do... logging objects is nasty in the new version of Winston, half tempted to force 2.x.

Gotta log like:

```
logger.main.debug(util.format('%o',settings))
```

I did drop those two log lines from the admin.js though, as they're a bit too noisy even for debug level logging.

**Comment by marshyonline** [comment]
Oh it is version 2 😆  disregard :P

**Comment by DanrwAU** [comment]
Just running through the plugins again - not sure what this bit is attempting to log but it looks abit funky - otherwise all good 👍  😉 

```
3|pagermon  | 2018-12-04 10:50 +00:00: [pmon]  2018-12-04 21:12:10  debug : { id: 153294,
3|pagermon  | 2018-12-04 10:50 +00:00:   address: '0123456',
3|pagermon  | 2018-12-04 10:50 +00:00:   message: 'Test Message 1',
3|pagermon  | 2018-12-04 10:50 +00:00:   source: 'Site 99',
3|pagermon  | 2018-12-04 10:50 +00:00:   timestamp: '2018-12-04 21:12:84',
3|pagermon  | 2018-12-04 10:50 +00:00:   alias_id: 225,
3|pagermon  | 2018-12-04 10:50 +00:00:   alias: 'Internal Test',
3|pagermon  | 2018-12-04 10:50 +00:00:   agency: 'SYSTEM',
3|pagermon  | 2018-12-04 10:50 +00:00:   icon: 'wifi',
3|pagermon  | 2018-12-04 10:50 +00:00:   color: 'green',
3|pagermon  | 2018-12-04 10:50 +00:00:   ignore: 0,
3|pagermon  | 2018-12-04 10:50 +00:00:   aliasMatch: 225,
3|pagermon  | 2018-12-04 10:50 +00:00:   pluginconf:
3|pagermon  | 2018-12-04 10:50 +00:00:    { Discord: { enable: false, webhook: '' },
3|pagermon  | 2018-12-04 10:50 +00:00:      Pushover:
3|pagermon  | 2018-12-04 10:50 +00:00:       { enable: true,
3|pagermon  | 2018-12-04 10:50 +00:00:         priority: { value: '0' },
3|pagermon  | 2018-12-04 10:50 +00:00:         group: 'HAHA NOT THIS TIME GITHUB',
3|pagermon  | 2018-12-04 10:50 +00:00:         sound: { value: '' } },
3|pagermon  | 2018-12-04 10:50 +00:00:      SMTP: { enable: false, mailto: '' },
3|pagermon  | 2018-12-04 10:50 +00:00:      Telegram: { enable: false, chat: '' },
3|pagermon  | 2018-12-04 10:50 +00:00:      Twitter: { enable: false, hashtag: '' },
3|pagermon  | 2018-12-04 10:50 +00:00:      Filter: {},
3|pagermon  | 2018-12-04 10:50 +00:00:      MicrosoftTeams: {},
3|pagermon  | 2018-12-04 10:50 +00:00:      Slack: {},
3|pagermon  | 2018-12-04 10:50 +00:00:      Template: {} },
3|pagermon  | 2018-12-04 10:50 +00:00:   pluginData: {},
3|pagermon  | 2018-12-04 10:50 +00:00:   level: '\u001b[34mdebug\u001b[39m',
3|pagermon  | 2018-12-04 10:50 +00:00:   label: '[pmon]',
3|pagermon  | 2018-12-04 10:50 +00:00:   [Symbol(level)]: 'debug',
3|pagermon  | 2018-12-04 10:50 +00:00:   [Symbol(message)]:
3|pagermon  | 2018-12-04 10:50 +00:00:    '[pmon]  2018-12-04 21:12:84  \u001b[34mdebug\u001b[39m : Test Message 1' }

```

**Comment by davidmckenzie** [comment]
Merged in other stuff, now merging \o/

**Pull request merged**